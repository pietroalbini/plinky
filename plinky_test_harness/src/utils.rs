use anyhow::{bail, Error};
use std::path::{Path, PathBuf};
use std::process::Command;

pub fn run(command: &mut Command) -> Result<(), Error> {
    let cmd_repr = format!("{command:?}");
    let output = command.output()?;
    if !output.status.success() {
        eprintln!("Failed to execute {cmd_repr}");
        eprintln!();
        eprintln!("=== stdout ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stdout));
        eprintln!();
        eprintln!("=== stderr ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stderr));
        eprintln!();
        bail!("command failed with exit {}", output.status);
    }
    Ok(())
}

pub struct RunAndSnapshot {
    name: String,
    path: PathBuf,
    output: String,
}

impl RunAndSnapshot {
    pub fn new(name: &str, path: &Path) -> Self {
        Self { name: name.into(), path: path.into(), output: String::new() }
    }

    pub fn run(&mut self, action: &str, command: &mut Command) -> Result<bool, Error> {
        self.separator();
        match command.output() {
            Ok(output) => {
                self.output.push_str(&format!("{action} exited with {}\n", output.status));
                for (name, content) in [("stdout", &output.stdout), ("stderr", &output.stderr)] {
                    if content.trim_ascii().is_empty() {
                        self.output.push_str(&format!("\nno {name} present\n"));
                    } else {
                        let mut content = String::from_utf8_lossy(content).to_string();
                        content = content.replace(env!("CARGO_MANIFEST_DIR"), "${project}");

                        self.output
                            .push_str(&format!("\n=== {name} ===\n{}\n", content.trim_ascii_end()));
                    }
                }
                Ok(output.status.success())
            }
            Err(err) => {
                self.output.push_str(&format!("{action} failed to execute with error: {err}"));
                Ok(false)
            }
        }
    }

    pub fn note(&mut self, note: &str) {
        self.separator();
        self.output.push_str(note);
        if !note.ends_with('\n') {
            self.output.push('\n');
        }
    }

    fn separator(&mut self) {
        if !self.output.is_empty() {
            self.output.push_str("\n==============\n\n");
        }
    }

    pub fn persist(self) {
        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_prepend_module_to_snapshot(false);
        insta_settings.set_omit_expression(true);
        insta_settings.set_snapshot_path(self.path.canonicalize().unwrap());

        insta_settings.bind(|| {
            insta::assert_snapshot!(self.name, self.output.trim_ascii_end());
        });
    }
}

pub(crate) fn file_name(path: impl AsRef<Path>) -> String {
    path.as_ref().file_name().unwrap().to_str().unwrap().to_string()
}

#[track_caller]
pub(crate) fn err_str<T>(result: Result<T, Error>) -> Result<T, String> {
    match result {
        Ok(ok) => Ok(ok),
        Err(err) => {
            let mut repr = format!("error: {err}\n");
            let mut source = err.source();
            while let Some(err) = source {
                repr.push_str(&format!("  cause: {err}\n"));
                source = err.source();
            }
            panic!("{repr}")
        }
    }
}
