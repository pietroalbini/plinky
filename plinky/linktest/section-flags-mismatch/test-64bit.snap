---
source: plinky_test_harness/src/utils.rs
---
linking exited with exit status: 0

no stdout present

=== stderr ===
debug print: calculated layout
 │
 │    Layout:
 │  ╭────────────────────┬─────────────┬─────────────┬────────────────┬───────────────╮
 │  │ Part               │ File offset │ File length │ Memory address │ Memory length │
 │  ├────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <elf header>       │ 0x0         │ 0x40        │ -              │ -             │
 │  │ <program header>   │ 0x40        │ 0x118       │ -              │ -             │
 │  │ <section header>   │ 0x158       │ 0x200       │ -              │ -             │
 │  │ .shstrtab          │ 0x358       │ 0x4b        │ -              │ -             │
 │  │ .comment           │ 0x3a3       │ 0xf         │ -              │ -             │
 │  │ .strtab            │ 0x3b2       │ 0x9         │ -              │ -             │
 │  │ .symtab            │ 0x3bb       │ 0x30        │ -              │ -             │
 │  │ <padding>          │ 0x3eb       │ 0xc15       │ -              │ -             │
 │  │ .text.foo#0        │ 0x1000      │ 0x1         │ 0x400000       │ 0x1           │
 │  │ <padding>          │ 0x1001      │ 0xfff       │ -              │ -             │
 │  │ .text.foo#1        │ 0x2000      │ 0x1         │ 0x401000       │ 0x1           │
 │  │ <padding>          │ 0x2001      │ 0x7         │ -              │ -             │
 │  │ .note.gnu.property │ 0x2008      │ 0x30        │ 0x401008       │ 0x30          │
 │  ╰────────────────────┴─────────────┴─────────────┴────────────────┴───────────────╯
 │
 │    Segments:
 │  ╭──────────┬────────┬──────────────┬───────┬────────────────────╮
 │  │ Start    │ Align  │ Type         │ Perms │ Content            │
 │  ├──────────┼────────┼──────────────┼───────┼────────────────────┤
 │  │ 0x400000 │ 0x1000 │ program      │ RX    │ .text.foo#0        │
 │  │ 0x401000 │ 0x1000 │ program      │ RWX   │ .text.foo#1        │
 │  │ -        │ 0x1    │ GNU stack    │ RW    │                    │
 │  │ 0x401008 │ 0x8    │ GNU property │ R     │ .note.gnu.property │
 │  │ 0x401008 │ 0x8    │ notes        │ R     │ .note.gnu.property │
 │  ╰──────────┴────────┴──────────────┴───────┴────────────────────╯
 ┴
